parking_lot/parking_lot.py

```diff
@@ -1,8 +1,150 @@
 # -*- coding: utf-8 -*-
-from flask import Flask
+import datetime
 
-app = Flask(__name__)
 
-@app.route("/")
-def hello_world():
-    return "<p>Hello, World!</p>"
+class Parking_lot_error( Exception ):
+    message = 'unhandle error with parking lot'
+
+    def __init__( self, message=None ):
+        if message is None:
+            message = self.message
+        self.data = {
+            'status': 'error',
+            'error': message
+        }
+
+
+class Full_parking_error( Parking_lot_error ):
+    message = 'No free space.'
+
+
+class Cannot_find_car_error( Parking_lot_error ):
+    message = 'Cannot find a car in the location {}.'
+
+    def __init__( self, location=None ):
+        message = self.message.format( location )
+        self.data = {
+            'status': 'error',
+            'error': message
+        }
+
+
+class Car:
+    def __init__(
+            self, car, tariff, location, hourly_fee, daily_fee,
+            start=None, finish=None ):
+        self.car = car
+        self.tariff = tariff
+        self.location = location
+        if start is None:
+            start = datetime.datetime.now()
+        self.start = start
+        self.finish = finish
+
+        self.hourly_fee = hourly_fee
+        self.daily_fee = daily_fee
+
+    def exit( self ):
+        self.finish = datetime.datetime.now()
+
+    @property
+    def fee( self ):
+        if self.is_free_charge:
+            return 0
+
+        if self.is_daily:
+            return self.proportional_diff_time * self.daily_fee
+        if self.is_hourly:
+            return self.proportional_diff_time * self.hourly_fee
+        raise NotImplementedError(
+            f"the tariff {self.tariff} is not implemented" )
+
+    @property
+    def is_free_charge( self ):
+        diff_time = self.finish - self.start
+        return diff_time.total_seconds() // ( 60 * 15 ) < 1
+
+    @property
+    def is_daily( self ):
+        return self.tariff == 'daily'
+
+    @property
+    def is_hourly( self ):
+        return self.tariff == 'hourly'
+
+    @property
+    def proportional_diff_time( self ):
+        diff_time = self.finish - self.start
+        if self.is_hourly:
+            return ( diff_time.total_seconds() // 3600 ) + 1
+        if self.is_daily:
+            return ( diff_time.total_seconds() // 86400 ) + 1
+
+
+class Parking_lot:
+    locations = None
+    lot = None
+
+    def __init__( self, amount, hourly_fee=1, daily_fee=20 ):
+        self.lot = {}
+        self.locations = []
+        self.amount_lot = amount
+        self.hourly_fee = hourly_fee
+        self.daily_fee = daily_fee
+
+    @property
+    def amount_lot( self ):
+        return self._amount_lot
+
+    @amount_lot.setter
+    def amount_lot( self, value ):
+        self._amount_lot = value
+        self.locations = [ None ] * self._amount_lot
+        for v in self.lot.values():
```

Comment:
what is this code intended for ? is the purpose to adjust size of the parking lot on runtime?

Reply:
yes, is in the case of the edge case if need to change in the rest api and was helpful in the tests